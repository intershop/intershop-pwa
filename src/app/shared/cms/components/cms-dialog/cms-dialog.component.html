<!-- dialog content can either be in the content parameter or in the slot -->
<div
  *ngIf="pagelet && pagelet.stringParam('Content')"
  [ishServerHtml]="pagelet.stringParam('Content')"
  data-testing-id="dialog-content"
></div>

<ish-content-slot [slot]="'app_sf_base_cm:slot.dialog.content.pagelet2-Slot'" [pagelet]="pagelet"></ish-content-slot>
