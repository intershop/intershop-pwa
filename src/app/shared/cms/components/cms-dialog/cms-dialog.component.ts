import { ChangeDetectionStrategy, Component, Input } from '@angular/core';

import { ContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { CMSComponent } from 'ish-shared/cms/models/cms-component/cms-component.model';

@Component({
  selector: 'ish-cms-dialog',
  templateUrl: './cms-dialog.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CMSDialogComponent implements CMSComponent {
  @Input() pagelet: ContentPageletView;
}
