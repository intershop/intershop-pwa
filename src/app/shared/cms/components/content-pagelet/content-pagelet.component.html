<ish-content-design-view-wrapper [pageletId]="pageletId">
  <ng-container #cmsOutlet />
</ish-content-design-view-wrapper>
