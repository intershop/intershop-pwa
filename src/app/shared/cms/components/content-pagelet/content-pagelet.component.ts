import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ComponentFactoryResolver,
  Injector,
  Input,
  OnChanges,
  OnDestroy,
  OnInit,
  SimpleChange,
  ViewChild,
  ViewContainerRef,
} from '@angular/core';
import { ReplaySubject, Subject } from 'rxjs';
import { switchMap, takeUntil } from 'rxjs/operators';

import { CMSFacade } from 'ish-core/facades/cms.facade';
import { ContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { whenTruthy } from 'ish-core/utils/operators';
import { CMSComponentProvider, CMS_COMPONENT } from 'ish-shared/cms/configurations/injection-keys';
import { CMSComponent } from 'ish-shared/cms/models/cms-component/cms-component.model';
import { SfeAdapterService } from 'ish-shared/cms/sfe-adapter/sfe-adapter.service';
import { SfeMetadataWrapper } from 'ish-shared/cms/sfe-adapter/sfe-metadata-wrapper';
import { SfeMapper } from 'ish-shared/cms/sfe-adapter/sfe.mapper';

/**
 * The Content Pagelet Container Component renders the pagelet for the given 'pageletId'.
 * For the rendering an Angular component is used that has to be provided
 * for the DefinitionQualifiedName of the pagelet.
 *
 * @example
 * <ish-content-pagelet [pageletId]="pagelet"></ish-content-pagelet>
 */
@Component({
  selector: 'ish-content-pagelet',
  templateUrl: './content-pagelet.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ContentPageletComponent extends SfeMetadataWrapper implements OnChanges, OnInit, OnDestroy {
  /**
   * The Id of the Pagelet that is to be rendered.
   */
  @Input() pageletId: string;

  @ViewChild('cmsOutlet', { read: ViewContainerRef, static: true }) cmsOutlet: ViewContainerRef;

  private pageletId$ = new ReplaySubject<string>(1);
  private destroy$ = new Subject();

  constructor(
    private injector: Injector,
    private componentFactoryResolver: ComponentFactoryResolver,
    private sfeAdapter: SfeAdapterService,
    private cmsFacade: CMSFacade,
    private cdRef: ChangeDetectorRef
  ) {
    super();
  }

  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }

  ngOnInit() {
    this.pageletId$
      .pipe(
        switchMap(pageletId => this.cmsFacade.pagelet$(pageletId)),
        whenTruthy(),
        takeUntil(this.destroy$)
      )
      .subscribe(pagelet => {
        this.mapComponent(pagelet);
        if (this.sfeAdapter.isInitialized()) {
          this.setSfeMetadata(SfeMapper.mapPageletViewToSfeMetadata(pagelet));
        }
        this.cdRef.markForCheck();
      });
  }

  ngOnChanges() {
    this.pageletId$.next(this.pageletId);
  }

  private mapComponent(pagelet: ContentPageletView) {
    const components = this.injector.get<CMSComponentProvider[]>(CMS_COMPONENT, []);
    const mappedComponent = components.find(c => c.definitionQualifiedName === pagelet.definitionQualifiedName);

    if (mappedComponent) {
      const componentRef = this.createComponent(mappedComponent);
      this.initializeComponent(componentRef.instance, pagelet);
    } else {
      console.warn(`did not find mapping for ${pagelet.id} (${pagelet.definitionQualifiedName})`);
    }
  }

  private createComponent(mappedComponent: CMSComponentProvider) {
    const factory = this.componentFactoryResolver.resolveComponentFactory(mappedComponent.class);
    this.cmsOutlet.clear();
    return this.cmsOutlet.createComponent(factory);
  }

  private initializeComponent(instance: CMSComponent, pagelet: ContentPageletView) {
    instance.pagelet = pagelet;

    // OnChanges has to be manually invoked on dynamically created components
    if (instance.ngOnChanges) {
      instance.ngOnChanges({ pagelet: new SimpleChange(undefined, pagelet, true) });
    }
  }
}
