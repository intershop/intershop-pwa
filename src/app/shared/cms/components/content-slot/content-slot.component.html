<ng-content *ngIf="wrapper; else renderPagelets"></ng-content>

<ng-template #renderPagelets>
  <ish-content-pagelet
    *ngFor="let pagelet of pagelet.slot(slot).pageletIDs"
    [pageletId]="pagelet"
    [callParameters]="callParameters"
  ></ish-content-pagelet>
</ng-template>
