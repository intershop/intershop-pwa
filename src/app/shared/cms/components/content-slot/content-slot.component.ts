import { ChangeDetectionStrategy, Component, Input, OnChanges } from '@angular/core';

import { ContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { SfeAdapterService } from 'ish-shared/cms/sfe-adapter/sfe-adapter.service';
import { SfeMetadataWrapper } from 'ish-shared/cms/sfe-adapter/sfe-metadata-wrapper';
import { SfeMapper } from 'ish-shared/cms/sfe-adapter/sfe.mapper';

/**
 * The Content Slot Container Component renders the assigned sub pagelets
 * of the identified 'slot' of the given 'pagelet'.
 * By default it is just using the {@link ContentPageletContainerComponent}
 * to render each sub pagelet.
 * If more specific HTML or functionality is needed for the rendering
 * the 'wrapper' flag needs to be used to use the provided HTML for rendering.
 *
 * @example
 * <ish-content-slot [slot]="'app_pwa:slot.pagelet2-Slot'" [pagelet]="pagelet"></ish-content-slot>
 *
 * <ish-content-slot [wrapper]="true" [slot]="'app_pwa:slot.pagelet2-Slot'" [pagelet]="pagelet">
 *               <div *ngFor="let slotPagelet of slotPagelets">
 *                 <ish-content-pagelet [pageletId]="slotPagelet"></ish-content-pagelet>
 *               </div>
 * </ish-content-slot>
 */
@Component({
  selector: 'ish-content-slot',
  templateUrl: './content-slot.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ContentSlotComponent extends SfeMetadataWrapper implements OnChanges {
  /**
   * The DefinitionQualifiedName of the slot that should be rendered.
   */
  @Input() slot: string;
  /**
   * The parent Pagelet that contains the slot.
   */
  @Input() pagelet: ContentPageletView;
  /**
   * An optional flag that controls the rendering of the pagelets with the wrapped HTML content.
   */
  @Input() wrapper?: boolean;

  constructor(private sfeAdapter: SfeAdapterService) {
    super();
  }

  ngOnChanges() {
    if (this.pagelet && this.sfeAdapter.isInitialized()) {
      this.setSfeMetadata(SfeMapper.mapSlotViewToSfeMetadata(this.pagelet, this.slot));
    }
  }
}
