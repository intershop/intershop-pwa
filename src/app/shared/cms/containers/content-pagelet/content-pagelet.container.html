<ng-container #cmsOutlet></ng-container>
