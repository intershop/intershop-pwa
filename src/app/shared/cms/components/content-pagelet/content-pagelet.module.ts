import { Injector, NgModule } from '@angular/core';

import { ModuleLoaderService } from 'ish-core/utils/module-loader/module-loader.service';
import { ContentPageletComponent } from 'ish-shared/cms/components/content-pagelet/content-pagelet.component';

const exportedComponents = [ContentPageletComponent];

@NgModule({
  declarations: [...exportedComponents],
  exports: [...exportedComponents],
})
export class ContentPageletModule {
  constructor(moduleLoader: ModuleLoaderService, injector: Injector) {
    moduleLoader.init(injector);
  }
}
