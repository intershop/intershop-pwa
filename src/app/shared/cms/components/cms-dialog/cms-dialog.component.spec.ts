import { APP_BASE_HREF } from '@angular/common';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { MockComponent } from 'ng-mocks';
import { instance, mock, when } from 'ts-mockito';

import { ServerHtmlDirective } from 'ish-core/directives/server-html.directive';
import { AppFacade } from 'ish-core/facades/app.facade';
import { ContentPagelet } from 'ish-core/models/content-pagelet/content-pagelet.model';
import { createContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { ContentSlotComponent } from 'ish-shared/cms/components/content-slot/content-slot.component';

import { CMSDialogComponent } from './cms-dialog.component';

describe('Cms Dialog Component', () => {
  let component: CMSDialogComponent;
  let fixture: ComponentFixture<CMSDialogComponent>;
  let element: HTMLElement;

  beforeEach(async () => {
    const appFacade = mock(AppFacade);
    when(appFacade.icmBaseUrl).thenReturn('http://example.com');

    await TestBed.configureTestingModule({
      imports: [RouterTestingModule],
      declarations: [CMSDialogComponent, MockComponent(ContentSlotComponent), ServerHtmlDirective],
      providers: [
        { provide: AppFacade, useFactory: () => instance(appFacade) },
        { provide: APP_BASE_HREF, useValue: '/' },
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(CMSDialogComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => fixture.detectChanges()).not.toThrow();
  });

  it('should be display pagelet content if content variable is not empty', () => {
    const pagelet = {
      domain: 'domain',
      displayName: 'pagelet1',
      definitionQualifiedName: 'fq',
      id: 'id',
      configurationParameters: { Content: 'testContent' },
    } as ContentPagelet;
    const pageletView = createContentPageletView(pagelet);
    component.pagelet = pageletView;
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element.querySelector('[data-testing-id="dialog-content"]').innerHTML).toBe('testContent');
  });

  it('should be display slot content if slot is not empty', () => {
    const pagelet2 = {
      domain: 'domain',
      displayName: 'pagelet1',
      definitionQualifiedName: 'fq',
      id: 'id',
      configurationParameters: {},
      slots: [
        {
          displayName: 'dialog',
          definitionQualifiedName: 'app_sf_base_cm:slot.dialog.content.pagelet2-Slot',
          pageletIDs: ['dialogId'],
        },
      ],
    };

    component.pagelet = createContentPageletView(pagelet2);

    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchInlineSnapshot(
      `<ish-content-slot ng-reflect-slot="app_sf_base_cm:slot.dialog.con"></ish-content-slot>`
    );
  });
});
