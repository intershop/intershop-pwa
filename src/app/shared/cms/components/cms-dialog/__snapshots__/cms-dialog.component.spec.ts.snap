// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cms Dialog Component should be display slot content if slot is not empty 1`] = `
<ish-content-slot
  ng-reflect-slot="app_sf_responsive_cm:slot.dial"
  ng-reflect-pagelet="[object Object]"
></ish-content-slot>
`;
